### PR TITLE
feat(levm): implement precompile blake2f

### PR DESCRIPTION
**Pull request by maximopalopoli**
**Motivation**

The goal is to implement the blake2f precompile.

**Description**

The blake compressing functionality is based on [RFC 7693](https://datatracker.ietf.org/doc/html/rfc7693).

There are some related EF tests not passing, I left a commented test that simulate one of those cases. The test cases that do not pass are from `stPreCompiledContracts/blake2B.json`.

### CODE REVIEW

**Comment by JereSalo** [comment]
Have you tried using a rust crate like [Blake2](https://docs.rs/blake2/latest/blake2/). Couldn't that get some of the complexity out of the way here?

File: crates/vm/levm/src/precompiles.rs [hunk 1]
```diff
@@ -678,12 +680,211 @@ fn update_pairing_result(
...
+    calldata: &Bytes,
+    gas_for_call: u64,
+    consumed_gas: &mut u64,
 ) -> Result<Bytes, VMError> {
-    Ok(Bytes::new())
+    if calldata.len() != 213 {
+        return Err(VMError::PrecompileError(PrecompileError::ParsingInputError));
+    }
+
+    let rounds: U256 = calldata
```
Comment (JereSalo) [inline crates/vm/levm/src/precompiles.rs:806]: why `U256` for a 4 byte value?

**Comment by JereSalo** [comment]
I feel like some of the errors propagated should actually be internal errors if they are not intended to happen independently of the input received.

**Comment by maximopalopoli** [comment]
> Have you tried using a rust crate like [Blake2](https://docs.rs/blake2/latest/blake2/). Couldn't that get some of the complexity out of the way here?

I tried to use related crates, but I couldn't find one that does what we specifically need. The functionality is similar to the functionalities that are related to the Blake2 algorithm, but this one is more configurable than the others, since it receives parameters that the others do not.

**Review by JereSalo — COMMENTED** [review]
I think that comments can be added for more clarity.
Can we abstract the logic for parsing the input into one function? (If you think it is worth it, otherwise we can leave comments but it will be easier to read if we do)

File: crates/vm/levm/src/precompiles.rs (hunk 1 above)
Comment (maximopalopoli) [reply crates/vm/levm/src/precompiles.rs:806]: It's an intermediate size to then move on to a usize, I couldn't find a prettier way to do it.

File: crates/vm/levm/tests/tests.rs [hunk 2]
```diff
@@ -4643,3 +4646,50 @@ fn ecpairing_test() {
         (ECPAIRING_BASE_COST + ECPAIRING_GROUP_COST * 2)
     );
 }
+
+#[test]
+fn blake2f_test() {
```
Comment (lima-limon-inc) [inline crates/vm/levm/tests/tests.rs:4651]: Maxi, what do you think about adding a source for these tests? Just in the case in the future it's easier to find them

Something like: 
```suggestion
#[test]
// Source: https://eips.ethereum.org/EIPS/eip-152#test-vector-1
fn blake2f_test() {
```

File: crates/vm/levm/tests/tests.rs [hunk 3]
```diff
@@ -4643,3 +4646,50 @@ fn ecpairing_test() {
...
+
+    let expected_result = Bytes::from(hex::decode("ba80a53f981c4d0d6a2797b69f12f6e94c212f14685ac4b74b12bb6fdbffa2d17d87c5392aab792dc252d5de4533cc9518d38aa8dbf1925ab92386edd4009923").unwrap());
+
+    assert_eq!(result, expected_result);
+    assert_eq!(consumed_gas, 12 * BLAKE2F_ROUND_COST);
+}
+
+/*
+#[test]
+fn blake2f_test_2() {
```
Comment (lima-limon-inc) [inline crates/vm/levm/tests/tests.rs:4666]: Ditto
```suggestion
#[test]
// Source: https://eips.ethereum.org/EIPS/eip-152#test-vector-2
fn blake2f_test_2() {
```

File: crates/vm/levm/src/precompiles.rs [hunk 4]
```diff
@@ -678,12 +682,206 @@ fn update_pairing_result(
...
+fn g(v: [u64; 16], a: usize, b: usize, c: usize, d: usize, x: u64, y: u64) -> [u64; 16] {
+    let mut ret = v;
+    ret[a] = v[a].wrapping_add(v[b]).wrapping_add(x);
+    ret[d] = (ret[d] ^ ret[a]).rotate_right(R1);
+    ret[c] = ret[c].wrapping_add(ret[d]);
+    ret[b] = (ret[b] ^ ret[c]).rotate_right(R2);
+    ret[a] = ret[a].wrapping_add(ret[b]).wrapping_add(y);
+    ret[d] = (ret[d] ^ ret[a]).rotate_right(R3);
+    ret[c] = ret[c].wrapping_add(ret[d]);
+    ret[b] = (ret[b] ^ ret[c]).rotate_right(R4);
```
Comment (lima-limon-inc) [inline crates/vm/levm/src/precompiles.rs:734]: @maximopalopoli Is it possible that you are not applying the modulus operator here?

File: crates/vm/levm/src/precompiles.rs [hunk 5]
```diff
@@ -678,12 +682,206 @@ fn update_pairing_result(
...
+    let mut h = [0; 8];
+    for i in 0..8_usize {
+        let data_read = read_bytes_from_offset(calldata, 4, i)?;
+
+        let read_slice = h.get_mut(i).ok_or(InternalError::SlicingError)?;
+        *read_slice = data_read;
+    }
+
+    let mut m = [0; 16];
+    for i in 0..8_usize {
```
Comment (damiramirez) [inline crates/vm/levm/src/precompiles.rs:831]: Don't we have to iterate up to 16 here?

File: crates/vm/levm/src/precompiles.rs (hunk 4 above)
Comment (lima-limon-inc) [reply crates/vm/levm/src/precompiles.rs:734]: See: https://datatracker.ietf.org/doc/html/rfc7693#section-3.1

File: crates/vm/levm/src/precompiles.rs (hunk 5 above)
Comment (lima-limon-inc) [reply crates/vm/levm/src/precompiles.rs:831]: @maximopalopoli I think this is it!

Comment (maximopalopoli) [reply crates/vm/levm/src/precompiles.rs]: Done in [Fix times iterating when reading m](https://github.com/lambdaclass/ethrex/pull/1563/commits/1c136696d1bd891a34b114f55c4b7a31e51297d5)!

File: crates/vm/levm/tests/tests.rs (hunk 2 above)
Comment (maximopalopoli) [reply crates/vm/levm/tests/tests.rs:4651]: Done in [Add source of first test](https://github.com/lambdaclass/ethrex/pull/1563/commits/42fa3c8eb596fd21cd8c1fc6602c25cd4de9ba59)

File: crates/vm/levm/tests/tests.rs (hunk 3 above)
Comment (maximopalopoli) [reply crates/vm/levm/tests/tests.rs:4666]: "Done" in [Add source of second test](https://github.com/lambdaclass/ethrex/pull/1563/commits/c5a6d26a06c5bbdba26494dc87f01e96935f11c1)

**Review by JereSalo — APPROVED** [review]
good job! it must have been hard to implement this

**Pull request merged**